### ISSUE TITLE
feat: embed, io/fs support

### ISSUE THREAD

**Issue by johejo** [golang-migrate/migrate#471]
**Is your feature request related to a problem? Please describe.**

The next Go 1.16 will add official file system interface and file embedding.

https://github.com/golang/go/issues/41190
https://github.com/golang/go/tree/master/src/io/fs

https://github.com/golang/go/issues/41191
https://github.com/golang/go/tree/master/src/embed

It would be great if golang-migrate supports these as source.Driver.

It replaces statik, go-bindata, and pkger.
Of course, this does not mean that they cannot be used.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

We will also need to use the build tag properly.

**Comment by dhui** [comment]
Thanks for sharing!

There's already support for a [httpfs](https://github.com/golang-migrate/migrate/tree/v4.13.0/source/httpfs) source (`net/http/FileSystem`), so we could do something similar for `io/fs/FS`.

I [believe](https://go.googlesource.com/proposal/+/master/design/draft-embed.md#file-api), if an `io/fs/FS` source driver existed, we wouldn't need an explicit `embed` driver.

We'll need to add a [build constraint](https://golang.org/cmd/go/#hdr-Build_constraints) for `go1.16` for the `io/fs/FS` source driver.

**Comment by johejo** [comment]
We will not be able to include this feature until Go 1.15 is no longer supported. #480 
However, some users may want to use this feature, so I created an extra module.
https://github.com/johejo/golang-migrate-extra
This is based on #472 .

**Comment by dhui** [comment]
Re-opening and will revert the [revert](https://github.com/golang-migrate/migrate/pull/481) when Go 1.15 is no longer supported by `migrate`

**Comment by wkoszek** [comment]
@dhui @johejo Any way we (segmed.ai) can help here? We can provide 2 pairs of eyes to help with getting this support integrated, and also could help with testing. @Fenges

**Comment by dhui** [comment]
@wkoszek There's nothing anyone can do but wait for Go 1.15 to be not be officially supported e.g. when the old supported version of Go (1.16) has the `io/fs` package. Otherwise, people running migrate on older versions of Go will run into this [issue](https://github.com/golang-migrate/migrate/issues/480)

**Comment by philippta** [comment]
Possible workaround until Go 1.15 is no longer supported:

```
main.go
migrations/
  1_create_tables.up.sql
  1_create_tables.down.sql
```

```go
import "github.com/golang-migrate/migrate/v4/source/httpfs"

//go:embed migrations
var migrations embed.FS

source, err := httpfs.New(http.FS(migrations), "migrations")
// migrate.NewInstance("httpfs", source, ...)
```

**Comment by prochac** [comment]
We should use build tags to support 1.16 features and keep the support for 1.15 at least until 1.17 is released
(Following https://golang.org/doc/devel/release.html#policy)

Take an example from here: 
https://github.com/pkg/errors/blob/master/go113.go

**Comment by dhui** [comment]
The [PR](https://github.com/golang-migrate/migrate/pull/472) that added support for `io/fs` had build constraints but modules ignore built constraints. See: https://github.com/golang-migrate/migrate/issues/480#issuecomment-731518493

**Comment by ivanduka** [comment]
> Possible workaround until Go 1.15 is no longer supported:
> 
> ```
> main.go
> migrations/
>   1_create_tables.up.sql
>   1_create_tables.down.sql
> ```
> 
> ```go
> import "github.com/golang-migrate/migrate/v4/source/httpfs"
> 
> //go:embed migrations
> var migrations embed.FS
> 
> source, err := httpfs.New(http.FS(migrations), "migrations")
> // migrate.NewInstance("httpfs", source, ...)
> ```

Thank you so much!
Small correction: we need to use `NewWithSourceInstance` instead of `NewInstance` in `migrate.NewInstance("httpfs", source, ...)`

**Comment by zikaeroh** [comment]
Go 1.16.2 and Go 1.15.10 (which were just released) include a fix for https://github.com/golang/go/issues/44557, meaning that io/fs and embed can be depended on and not break downstream users when they `go mod tidy` (so long as they update their tooling to these latest patch releases), so this should be unblocked whenever you feel like those patch releases have been out long enough.

**Comment by vtolstov** [comment]
any chance to get embed support in golang-migrate ?

**Comment by dhui** [comment]
Let's wait a bit more time (about a month or two) for the newer versions of Go to be adopted. Your other option is to fork this library and merge the PR into your fork.

**Comment by alovak** [comment]
https://groups.google.com/g/golang-announce/c/cu9SP4eSXMc?pli=1

> We have just released Go versions 1.16.4 and 1.15.12, minor point releases.
> This minor release includes a security fix according to the new security policy (#44918).

1.15.12 is already here!

P.S. just to support my previous comment :)

**Comment by philippta** [comment]
What about using build tag for this, that restricts the compilation to Go 1.16 and onwards?

**Comment by dhui** [comment]
Fixed by #560
This io/fs support is now available in the master branch but [won't be available in a release](https://github.com/golang-migrate/migrate/issues/546#issuecomment-823417672) until CI works for releases again
